### PR TITLE
Add hack script to aggregate traces in MC env

### PR DESCRIPTION
**Pull request by nrfox**
Aggregates tracing in local/testing multicluster environments.

### CODE REVIEW

**Comment by ScriptingShrimp** [comment]
> Aggregates tracing in local/testing multicluster environments.

just wondering why is good to have traces aggregated in the one of the instances

**Comment by nrfox** [comment]
> > Aggregates tracing in local/testing multicluster environments.
> 
> just wondering why is good to have traces aggregated in the one of the instances

Kiali only supports querying traces from a single endpoint. Same with metrics.

File: hack/istio/multicluster/setup-tracing.sh [hunk 1]
```diff
@@ -0,0 +1,121 @@
+#!/bin/bash
+
+set -e
+
+SCRIPT_DIR="$(cd $(dirname "${BASH_SOURCE[0]}") && pwd)"
+source ${SCRIPT_DIR}/env.sh $*
+
+# This scripts combines traces from the west cluster into the east cluster by sending traces through the istio-ingressgateway.
```
Comment (jmazzitelli) [inline hack/istio/multicluster/setup-tracing.sh:8]:
```suggestion
# This script combines traces from the west cluster into the east cluster by sending traces through the istio-ingressgateway.
```

File: hack/istio/multicluster/setup-tracing.sh [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+# It deploys an otel collector into the west cluster which gathers all traces and exports them to the east cluster.
+# It opens a separate port on the istio-ingressgateway for zipkin so that the tracing setup doesn't conflict with
+# the bookinfo gateway/traffic.
+
+# This adds a port to the istio-ingressgateway service. This could probably be done through istioctl and passing the right combination
+# of settings into the istioctl hack script but it's way simpler to just patch the service directly.
+ingress_output=$(kubectl get svc -n istio-system --context "${CLUSTER1_CONTEXT}" istio-ingressgateway -o jsonpath='{.spec.ports[?(@.name=="zipkin-http")]}')
```
Comment (jmazzitelli) [inline hack/istio/multicluster/setup-tracing.sh:15]:
```suggestion
ingress_output=$(${CLIENT_EXE} get svc -n istio-system --context "${CLUSTER1_CONTEXT}" istio-ingressgateway -o jsonpath='{.spec.ports[?(@.name=="zipkin-http")]}')
```

File: hack/istio/multicluster/setup-tracing.sh [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+# Check if the output is empty
+if [ -z "$ingress_output" ]; then
+    kubectl --context "${CLUSTER1_CONTEXT}" patch Service -n istio-system istio-ingressgateway --type=json -p '[{"op": "add", "path": "/spec/ports/-", "value": {"name": "zipkin-http", "port": 9411, "protocol": "TCP", "targetPort": 8080}}]'
```
Comment (jmazzitelli) [inline hack/istio/multicluster/setup-tracing.sh:18]:
```suggestion
    ${CLIENT_EXE} --context "${CLUSTER1_CONTEXT}" patch Service -n istio-system istio-ingressgateway --type=json -p '[{"op": "add", "path": "/spec/ports/-", "value": {"name": "zipkin-http", "port": 9411, "protocol": "TCP", "targetPort": 8080}}]'
```

File: hack/istio/multicluster/setup-tracing.sh [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+fi
+
+kubectl --context "${CLUSTER1_CONTEXT}" apply -f - <<EOF
```
Comment (jmazzitelli) [inline hack/istio/multicluster/setup-tracing.sh:21]:
```suggestion
${CLIENT_EXE} --context "${CLUSTER1_CONTEXT}" apply -f - <<EOF
```

File: hack/istio/multicluster/setup-tracing.sh [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  http:
+  - name: zipkin-route
+    route:
+    - destination:
+        host: zipkin.istio-system.svc.cluster.local
+        port:
+          number: 9411
+EOF
+
+kubectl --context "${CLUSTER1_CONTEXT}" apply -f - <<EOF
```
Comment (jmazzitelli) [inline hack/istio/multicluster/setup-tracing.sh:41]:
```suggestion
${CLIENT_EXE} --context "${CLUSTER1_CONTEXT}" apply -f - <<EOF
```

File: hack/istio/multicluster/setup-tracing.sh [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  servers:
+  - hosts:
+    - '*'
+    port:
+      name: http
+      number: 9411
+      protocol: HTTP
+EOF
+
+ISTIO_INGRESS_IP=$(kubectl --context "${CLUSTER1_CONTEXT}" get service -n istio-system istio-ingressgateway -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
```
Comment (jmazzitelli) [inline hack/istio/multicluster/setup-tracing.sh:59]:
```suggestion
ISTIO_INGRESS_IP=$(${CLIENT_EXE} --context "${CLUSTER1_CONTEXT}" get service -n istio-system istio-ingressgateway -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
```

File: hack/istio/multicluster/setup-tracing.sh [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    enabled: false
+  otlp:
+    enabled: false
+  otlp-http:
+    enabled: false
+EOF
+
+# Create a service for the collector named zipkin.istio-system to
+# make it consistent with the normal istio tracing service.
+kubectl --context "${CLUSTER2_CONTEXT}" apply -f - <<EOF
```
Comment (jmazzitelli) [inline hack/istio/multicluster/setup-tracing.sh:107]:
```suggestion
${CLIENT_EXE} --context "${CLUSTER2_CONTEXT}" apply -f - <<EOF
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
I didn't run it - but it looks mighty fine. Just change the hardcoded kubectl to using the env var like the rest of the scripts.

**Review by jmazzitelli — APPROVED** [review]

**Pull request merged**